File: c++/design_pattern/AbstractFactory.cpp
#include <iostream>
#include <map>
#include <cmath>
#include <initializer_list>

using namespace std;

/* 1. 추상 팩터리  */
	/*	- 하나의 객체가 아닌 연관된 여러종류의 객체들을 생성해야 할 경우	*/

namespace AbstractFactory{

	struct HotDrink{
		virtual ~HotDrink(){}
		virtual void prepare(int volume) = 0;
	};

	struct Tea : HotDrink{
		void prepare(int volume) override {
			std::cout << "Take tea bag, boil water, pour " << volume << "ml, add some lemon" << std::endl;
		}
	};

	struct Coffee : HotDrink{
		void prepare(int volume) override {
			std::cout << " Coffee" << volume << "ml" << std::endl;
		}
	};


	// Bad Case
	unique_ptr<HotDrink> make_drink(string type){
		unique_ptr<HotDrink> drink;
		if(type == "tea"){
			drink = make_unique<Tea>();
			drink->prepare(200);
		}else if( type =="coffee"){

			drink = make_unique<Coffee>();
			drink->prepare(50);
		}

		return drink;
	}

	//Good Case
	struct HotDrinkFactory{
		virtual ~HotDrinkFactory(){}
		virtual unique_ptr<HotDrink> make() const = 0;
	};

	struct CoffeeFactory : HotDrinkFactory{
		unique_ptr<HotDrink> make() const override{
			return make_unique<Coffee>();
		}
	};
	struct TeaFactory : HotDrinkFactory{
		unique_ptr<HotDrink> make() const override{
			return make_unique<Tea>();
		}
	};

	class DrinkFactory{
		map < string, unique_ptr<HotDrinkFactory> > hot_factories;

		public : 
			DrinkFactory(){
				hot_factories["coffee"] = make_unique<CoffeeFactory>();
				hot_factories["tea"] = make_unique<TeaFactory>();
			}

			unique_ptr<HotDrink> make_drink(const string& name ){
				auto drink = hot_factories[name]->make();
				drink->prepare(200);

				return drink;
			}
	};

	void do_abstract_factory(){
		DrinkFactory my_factory;
		auto my_drink = my_factory.make_drink("tea");
	}
}


/* 2. 함수형 팩터리  */
	/*	- 하나의 객체가 아닌 연관된 여러종류의 객체들을 생성해야 할 경우	*/


namespace FunctionalFactory{

	struct HotDrink{
		virtual ~HotDrink(){}
		virtual void prepare(int volume) = 0;
	};

	struct Tea : HotDrink{
		void prepare(int volume) override {
			std::cout << "Take tea bag, boil water, pour " << volume << "ml, add some lemon" << std::endl;
		}
	};

	struct Coffee : HotDrink{
		void prepare(int volume) override {
			std::cout << " Coffee" << volume << "ml" << std::endl;
		}
	};


	class DrinkWithVolumeFactory{
		map < string, function<unique_ptr<HotDrink>()> > factories;

		public : 
			DrinkWithVolumeFactory(){
				factories["tea"] = []{
					auto tea = make_unique<Tea>();
					tea->prepare(200);
					return tea;
				};

				factories["coffee"] =[]{
					auto coffee = make_unique<Coffee>();
					coffee->prepare(50);
					return coffee;
				};
			}
			inline unique_ptr<HotDrink> make_drink(const string& name){
				return factories[name]();
			}
	};


	void do_functional_factory(){
		DrinkWithVolumeFactory x;
		auto drink = x.make_drink("coffee");
	}
}


// 3. 추상팩터리 예제 2


namespace AbstractFactory2{

	struct Button {
		virtual void paint()=0;
	};
	struct CheckBox {
		virtual void paint()=0;
	};
	struct DropDown{
		virtual void paint()=0;
	};

	struct GUIFactory{
		virtual Button* createButton(){return nullptr;};
		virtual CheckBox* createCheckBox(){return nullptr;};
		virtual DropDown* createDropDown(){return nullptr;};
	};

	struct WinButton : Button{
		void paint() override{}
	};
	struct MacButton : Button{
		void paint() override{}
	};
	struct WinCheckBox : CheckBox{
		void paint() override{};
	};
	struct MacCheckBox : CheckBox{
		void paint() override{};
	};
	struct WinDropDown : DropDown{
		void paint() override{};
	};
	struct MacDropDown : DropDown{
		void paint() override{};
	};


	struct WinGUIFactory : GUIFactory{
		Button* createButton() override {return new WinButton();}
		CheckBox* createCheckBox() override {return new WinCheckBox();}
		DropDown* createDropDown() override {return new WinDropDown();}
	};
	struct MacGUIFactory : GUIFactory{
		Button* createButton() override {return new MacButton();}
		CheckBox* createCheckBox() override {return new MacCheckBox();}
		DropDown* createDropDown() override {return new MacDropDown();}
	};

	struct Application {
		
		GUIFactory guiFactory;
		Button* btn;
		CheckBox* chkbx;
		DropDown* drpdwn;

		explicit Application(GUIFactory guiFactory) : guiFactory(guiFactory){
			btn = guiFactory.createButton();
			chkbx = guiFactory.createCheckBox();
			drpdwn = guiFactory.createDropDown();
		};

	};

	void do_abstract_factory(){
		MacGUIFactory gui;
		Application app(gui);
	}

};



int main(void){

	AbstractFactory::do_abstract_factory();
	FunctionalFactory::do_functional_factory();
	AbstractFactory2::do_abstract_factory();

	return 0;
}

File: java/src/main/java/com/example/design_pattern/AbstractFactory.java
package com.example.design_pattern;


import java.util.*;
import java.util.function.*;

public class AbstractFactory {


	private static interface HotDrink {
		public abstract void prepare(int volume);
	}

	private static class Tea implements HotDrink {

		@Override
		public void prepare(int volume){
			System.out.println(String.format("Making Tea %sml...",volume));
		}
	}

	private static class Coffee implements HotDrink {
		
		@Override
		public void prepare(int volume){
			System.out.println(String.format("Making Coffee %sml...", volume));
		}

	}

	private static class AbstractFactory1{

		private static interface HotDrinkFactory{
			public HotDrink make();
		}

		private static class TeaFactory implements HotDrinkFactory{
			@Override 
			public HotDrink make(){
				return new Tea();
			}
		}

		private static class CoffeeFactory implements HotDrinkFactory {
			@Override
			public HotDrink make(){
				return new Coffee();
			}
		}

		private static class DrinkFactory {
			private static Map<String, HotDrinkFactory> hotFactories;

			public DrinkFactory(){
				hotFactories = new HashMap<>();
				hotFactories.put("tea", new TeaFactory());
				hotFactories.put("coffee", new CoffeeFactory());
			}

			public HotDrink makeDrink(String name){
				HotDrink drink = hotFactories.get(name).make();
				if( drink instanceof Tea){
					drink.prepare(50);
				}
				if( drink instanceof Coffee){
					drink.prepare(200);
				}
				return drink;
			}
		}

		public static void main(){
			DrinkFactory myFactory = new DrinkFactory();
			HotDrink myDrink = myFactory.makeDrink("tea");
		}
	}

	private static class FunctionalAbstractFactory{

		private static interface HotDrinkWithVolumeFactory{
			public HotDrink makeWithVolume();
		}

		private static class DrinkWithVolumeFactory {
			private Map<String, HotDrinkWithVolumeFactory> hotFactories;

			public DrinkWithVolumeFactory(){
				hotFactories = new HashMap<>();

				hotFactories.put("tea", () -> {
					HotDrink tea = new Tea();
					tea.prepare(50);
					return tea;
				});	
				hotFactories.put("coffee", ()-> {
					HotDrink coffee = new Coffee();
					coffee.prepare(200);
					return coffee;
				});
			}

			public HotDrink makeDrink(String name){
				return hotFactories.get(name).makeWithVolume();
			}
		}

		public static void main(){

			DrinkWithVolumeFactory myFactory = new DrinkWithVolumeFactory();
			HotDrink myDrink = myFactory.makeDrink("coffee");
			
		}
	}

	private static interface Button {
		abstract public void paint();
	}

	private static interface CheckBox {
		abstract public void paint();
	}

	private static interface DropDown {
		abstract public void paint();
	}

	private static class WinButton  implements Button{
		@Override 
		public void paint(){}
	}
	private static class MacButton  implements Button{
		@Override 
		public void paint(){}
	}

	private static class WinCheckBox  implements CheckBox{
		@Override 
		public void paint(){}
	}
	
	private static class MacCheckBox  implements CheckBox{
		@Override 
		public void paint(){}
	}

	private static class WinDropDown  implements DropDown{
		@Override 
		public void paint(){}
	}
	private static class MacDropDown  implements DropDown{
		@Override 
		public void paint(){}
	}



	private static class AbstractFactory2{


		private static interface GuiFactory {
			public Button createButton();
			public CheckBox createCheckBox();
			public DropDown createDropDown();
		}

		private static class WinGuiFactory implements GuiFactory{
			@Override 
			public Button createButton(){
				return new WinButton();
			}
			@Override 
			public CheckBox createCheckBox(){
				return new WinCheckBox();
			}
			@Override 
			public DropDown createDropDown(){
				return new WinDropDown();
			}
		}

		private static class MacGuiFactory implements GuiFactory{
			@Override 
			public Button createButton(){
				return new MacButton();
			}
			@Override 
			public CheckBox createCheckBox(){
				return new MacCheckBox();
			}
			@Override 
			public DropDown createDropDown(){
				return new MacDropDown();
			}
		}

		private static class Application {
			Button button;
			CheckBox checkBox;
			DropDown dropDown;
			Application(GuiFactory factory){
				button = factory.createButton();
				checkBox = factory.createCheckBox();
				dropDown = factory.createDropDown();
			}
		};

		public static void main(){

			GuiFactory factory = new MacGuiFactory();
			Application app = new Application(factory);
		}
	}


	public static void main(String... args){

		AbstractFactory1.main();
		FunctionalAbstractFactory.main();
		AbstractFactory2.main();
	}	
}